server/src/main/java/org/elasticsearch/action/bulk/TransportBulkAction.java

```diff
@@ -155,11 +157,14 @@ protected void doExecute(Task task, BulkRequest bulkRequest, ActionListener<Bulk
         ImmutableOpenMap<String, IndexMetaData> indicesMetaData = metaData.indices();
         for (DocWriteRequest<?> actionRequest : bulkRequest.requests) {
             IndexRequest indexRequest = getIndexWriteRequest(actionRequest);
+
             if (indexRequest != null) {
```

Comment:
The logic inside this if statement has grown a bit too much. I think we should extract this logic into a static method? We can then later also add unit tests. Something like this:

```
IndexRequest indexRequest = getIndexWriteRequest(actionRequest);
if (indexRequest != null) {
    boolean indexRequestHasPipeline = 
            resolveRequiredOrDefaultPipeline(indexRequest, localNodeIsIngestNode, metaData, indicesMetaData);
    if (indexRequestHasPipeline) {
        hasIndexRequestsWithPipelines = true;
    }
}

...

static boolean resolveRequiredOrDefaultPipeline(IndexRequest indexRequest,
                                                boolean localNodeIsIngestNode,
                                                MetaData metaData,
                                                ImmutableOpenMap<String, IndexMetaData> indicesMetaData) {
    if (indexRequest.isForwardedRequest() == false) {
        final String requestPipeline = indexRequest.getPipeline();
        indexRequest.setPipeline(IngestService.NOOP_PIPELINE_NAME);
        boolean requestCanOverridePipeline = true;
        String requiredPipeline = null;
        // start to look for default or required pipelines via settings found in the index meta data
        IndexMetaData indexMetaData = indicesMetaData.get(indexRequest.index());
        // check the alias for the index request (this is how normal index requests are modeled)
        if (indexMetaData == null && indexRequest.index() != null) {
            AliasOrIndex indexOrAlias = metaData.getAliasAndIndexLookup().get(indexRequest.index());
            if (indexOrAlias != null && indexOrAlias.isAlias()) {
                AliasOrIndex.Alias alias = (AliasOrIndex.Alias) indexOrAlias;
                indexMetaData = alias.getWriteIndex();
            }
        }
        // check the alias for the action request (this is how upserts are modeled)
        if (indexMetaData == null && indexRequest.index() != null) {
            AliasOrIndex indexOrAlias = metaData.getAliasAndIndexLookup().get(indexRequest.index());
            if (indexOrAlias != null && indexOrAlias.isAlias()) {
                AliasOrIndex.Alias alias = (AliasOrIndex.Alias) indexOrAlias;
                indexMetaData = alias.getWriteIndex();
            }
        }
        if (indexMetaData != null) {
            final Settings indexSettings = indexMetaData.getSettings();
            if (IndexSettings.REQUIRED_PIPELINE.exists(indexSettings)) {
                // find the required pipeline if one is defined from an existing index
                requiredPipeline = IndexSettings.REQUIRED_PIPELINE.get(indexSettings);
                assert IndexSettings.DEFAULT_PIPELINE.get(indexSettings).equals(IngestService.NOOP_PIPELINE_NAME) :
                    IndexSettings.DEFAULT_PIPELINE.get(indexSettings);
                indexRequest.setPipeline(requiredPipeline);
                requestCanOverridePipeline = false;
            } else {
                // find the default pipeline if one is defined from an existing index
                String defaultPipeline = IndexSettings.DEFAULT_PIPELINE.get(indexSettings);
                indexRequest.setPipeline(defaultPipeline);
            }
        } else if (indexRequest.index() != null) {
            // the index does not exist yet (and is valid request), so match index templates to look for a default pipeline
            List<IndexTemplateMetaData> templates = MetaDataIndexTemplateService.findTemplates(metaData, indexRequest.index());
            assert (templates != null);
            // order of templates are highest order first, we have to iterate through them all though
            String defaultPipeline = null;
            for (IndexTemplateMetaData template : templates) {
                final Settings settings = template.settings();
                if (requiredPipeline == null && IndexSettings.REQUIRED_PIPELINE.exists(settings)) {
                    requiredPipeline = IndexSettings.REQUIRED_PIPELINE.get(settings);
                    requestCanOverridePipeline = false;
                    // we can not break in case a lower-order template has a default pipeline that we need to reject
                } else if (defaultPipeline == null && IndexSettings.DEFAULT_PIPELINE.exists(settings)) {
                    defaultPipeline = IndexSettings.DEFAULT_PIPELINE.get(settings);
                    // we can not break in case a lower-order template has a required pipeline that we need to reject
                }
            }
            if (requiredPipeline != null && defaultPipeline != null) {
                // we can not have picked up a required and a default pipeline from applying templates
                final String message = String.format(
                    Locale.ROOT,
                    "required pipeline [%s] and default pipeline [%s] can not both be set",
                    requiredPipeline,
                    defaultPipeline);
                throw new IllegalArgumentException(message);
            }
            final String pipeline;
            if (requiredPipeline != null) {
                pipeline = requiredPipeline;
            } else {
                pipeline = Objects.requireNonNullElse(defaultPipeline, IngestService.NOOP_PIPELINE_NAME);
            }
            indexRequest.setPipeline(pipeline);
        }

        if (requestPipeline != null) {
            if (requestCanOverridePipeline == false) {
                final String message = String.format(
                    Locale.ROOT,
                    "request pipeline [%s] can not override required pipeline [%s]",
                    requestPipeline,
                    requiredPipeline);
                throw new IllegalArgumentException(message);
            } else {
                indexRequest.setPipeline(requestPipeline);
            }
        }

        if (localNodeIsIngestNode == false) {
            /*
                * We have to track whether or not the pipeline for this request has already been derived. It can happen that the
                * pipeline for this request has already been derived yet we execute this loop again. That occurs if the bulk
                * request has been forwarded by a non-ingest coordinating node to an ingest node. In this case, the coordinating
                * node will have already derived the pipeline for this request. It is important that we are able to distinguish
                * this situation as we can not double-derive the pipeline because we will not be able to distinguish the case of
                * the pipeline having been set from a request pipeline parameter versus having been set by the derivation. We need
                * to be able to distinguish these cases as we need to reject the request if the pipeline was set by a required
                * pipeline and there is a request pipeline parameter too.
                */
            indexRequest.isForwardedRequest(true);
        }
    }

    // Return whether this index request has a pipeline
    return IngestService.NOOP_PIPELINE_NAME.equals(indexRequest.getPipeline());
}
```

Reply:
Yes, I think this method is unruly too, but I didn't want to mix a complete refactoring of it with this PR. I can work on this in a follow-up.

Reply:
👍 Let's do that in a follow up and then also add the much needed unit tests. 
If you like then I can work on this task.

Reply:
@martijnvg If you'd like to pick up refactoring this, that would be great!

Reply:
I will pick it up this week.